File: app/src/main/java/com/example/homestay/ui/admin/AdminUsersActivity.java
package com.example.homestay.ui.admin;

import android.os.Bundle;
import android.text.*;
import android.util.Patterns;
import android.view.View;
import android.widget.*;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.*;
import com.example.homestay.HomestayApplication;
import com.example.homestay.R;
import com.example.homestay.data.entity.*;
import com.example.homestay.data.model.AdminUserData;
import com.example.homestay.data.repository.HomestayRepository;
import com.example.homestay.utils.*;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.checkbox.MaterialCheckBox;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import java.text.*;
import java.util.*;

public class AdminUsersActivity extends AppCompatActivity {
  private HomestayRepository repository;
  private AdminUserAdapter adapter;
  private ProgressBar progress;
  private final List<AdminUserData> users = new ArrayList<>();
  private Boolean lockedFilter;

  @Override
  protected void onCreate(Bundle s) {
    super.onCreate(s);
    setContentView(R.layout.activity_admin_users);
    SystemBarUtils.keepContentBelowStatusBar(this);
    repository = ((HomestayApplication) getApplication()).getRepository();
    Toolbar t = findViewById(R.id.toolbar);
    setSupportActionBar(t);
    if (getSupportActionBar() != null) {
      getSupportActionBar().setDisplayHomeAsUpEnabled(false);
      getSupportActionBar().setTitle("Quản lý người dùng");
    }
    setup();
    AdminNavigationUtils.setup(this, R.id.admin_nav_users);
    load();
  }

  private void setup() {
    progress = findViewById(R.id.progress_bar);
    adapter =
        new AdminUserAdapter(
            new ArrayList<>(), this::confirmDelete, this::details, this::edit);
    RecyclerView list = findViewById(R.id.rv_users);
    list.setLayoutManager(new LinearLayoutManager(this));
    list.setAdapter(adapter);
    TextInputEditText search = findViewById(R.id.et_admin_search);
    search.addTextChangedListener(
        new TextWatcher() {
          public void beforeTextChanged(CharSequence s, int a, int b, int c) {}

          public void onTextChanged(CharSequence s, int a, int b, int c) {
            filter(s.toString());
          }

          public void afterTextChanged(Editable e) {}
        });
    ((ChipGroup) findViewById(R.id.chip_user_status))
        .setOnCheckedStateChangeListener(
            (g, ids) -> {
              int id = ids.isEmpty() ? View.NO_ID : ids.get(0);
              lockedFilter =
                  id == R.id.chip_users_active
                      ? Boolean.FALSE
                      : id == R.id.chip_users_locked ? Boolean.TRUE : null;
              filter(text(search));
            });
  }

  private void load() {
    progress.setVisibility(View.VISIBLE);
    AppExecutors.io()
        .execute(
            () -> {
              List<Booking> all = repository.getAllBookingsNow();
              List<AdminUserData> result = new ArrayList<>();
              for (User u : repository.getAllUsersNow()) {
                int count = 0;
                double spent = 0;
                Long last = null;
                for (Booking b : all)
                  if (b.getUserId() == u.getId()) {
                    count++;
                    if ("completed".equals(b.getStatus())) spent += b.getTotalPrice();
                    if (last == null || b.getCreatedAt() > last) last = b.getCreatedAt();
                  }
                RateLimiter.AttemptStatus can = RateLimiter.canAttemptLogin(this, u.getEmail());
                int failed = RateLimiter.getFailedAttempts(this, u.getEmail());
                long seconds = RateLimiter.getLockedSecondsRemaining(this, u.getEmail());
                result.add(
                    new AdminUserData(
                        String.valueOf(u.getId()),
                        u.getEmail(),
                        u.getPhone(),
                        u.getFullName(),
                        u.getCreatedAt(),
                        failed,
                        !can.allowed,
                        !can.allowed && seconds > 365L * 24 * 3600,
                        can.lockedUntil,
                        seconds,
                        count,
                        spent,
                        last));
              }
              runOnUiThread(
                  () -> {
                    users.clear();
                    users.addAll(result);
                    filter(text((TextInputEditText) findViewById(R.id.et_admin_search)));
                    progress.setVisibility(View.GONE);
                  });
            });
  }

  private void filter(String q) {
    if (adapter == null) return;
    String key = q.trim().toLowerCase(Locale.ROOT);
    List<AdminUserData> out = new ArrayList<>();
    for (AdminUserData u : users)
      if ((lockedFilter == null || Boolean.TRUE.equals(u.getLocked()) == lockedFilter)
          && ((u.getFullName() + " " + u.getEmail() + " " + u.getPhone())
              .toLowerCase(Locale.ROOT)
              .contains(key))) out.add(u);
    adapter.updateUsers(out);
    findViewById(R.id.tv_empty).setVisibility(out.isEmpty() ? View.VISIBLE : View.GONE);
  }

  private void details(AdminUserData u) {
    AppExecutors.io()
        .execute(
            () -> {
              Map<Long, Room> rooms = new HashMap<>();
              for (Room r : repository.getAllRoomsNow()) rooms.put(r.getId(), r);
              List<Booking> history = new ArrayList<>();
              for (Booking b : repository.getAllBookingsNow())
                if (String.valueOf(b.getUserId()).equals(u.getId())) history.add(b);
              history.sort((a, b) -> Long.compare(b.getCreatedAt(), a.getCreatedAt()));
              SimpleDateFormat date = new SimpleDateFormat("dd/MM/yyyy", new Locale("vi", "VN"));
              StringBuilder h = new StringBuilder();
              for (int i = 0; i < Math.min(10, history.size()); i++) {
                Booking b = history.get(i);
                Room room = rooms.get(b.getRoomId());
                h.append("• ")
                    .append(room == null ? "Phòng #" + b.getRoomId() : room.getName())
                    .append(" — ")
                    .append(date.format(new Date(b.getCheckInDate())))
                    .append(" — ")
                    .append(b.getStatus())
                    .append('\n');
              }
              if (h.length() == 0) h.append("Chưa có lịch sử đặt phòng");
              String spent =
                  NumberFormat.getNumberInstance(new Locale("vi", "VN"))
                      .format((long) u.getTotalSpent());
              runOnUiThread(
                  () ->
                      new MaterialAlertDialogBuilder(this)
                          .setTitle(u.getFullName())
                          .setMessage(
                              "Mã người dùng: "
                                  + u.getId()
                                  + "\nEmail: "
                                  + u.getEmail()
                                  + "\nĐiện thoại: "
                                  + u.getPhone()
                                  + "\nNgày tham gia: "
                                  + date.format(new Date(u.getCreatedAt()))
                                  + "\nTrạng thái: "
                                  + (Boolean.TRUE.equals(u.getLocked()) ? "Bị khóa" : "Hoạt động")
                                  + "\nTổng booking: "
                                  + u.getBookingCount()
                                  + "\nTổng đã chi: "
                                  + spent
                                  + " đ\n\nLịch sử gần đây:\n"
                                  + h)
                          .setPositiveButton("Đóng", null)
                          .show());
            });
  }

  private void edit(AdminUserData data) {
    if (data.getEmail().equalsIgnoreCase(AdminAuth.EMAIL)) {
      toast("Không thể chỉnh sửa tài khoản quản trị hệ thống");
      return;
    }
    View v = getLayoutInflater().inflate(R.layout.dialog_admin_edit_user, null);
    TextInputEditText name = v.findViewById(R.id.et_admin_user_name),
        email = v.findViewById(R.id.et_admin_user_email),
        phone = v.findViewById(R.id.et_admin_user_phone),
        pass = v.findViewById(R.id.et_admin_user_password);
    MaterialCheckBox locked = v.findViewById(R.id.cb_admin_user_locked);
    name.setText(data.getFullName());
    email.setText(data.getEmail());
    phone.setText(data.getPhone());
    locked.setChecked(Boolean.TRUE.equals(data.getLocked()));
    androidx.appcompat.app.AlertDialog dialog =
        new MaterialAlertDialogBuilder(this).setView(v).create();
    v.findViewById(R.id.btn_cancel).setOnClickListener(x -> dialog.dismiss());
    v.findViewById(R.id.btn_save)
        .setOnClickListener(
            x -> {
              String n = text(name).trim(),
                  e = text(email).trim().toLowerCase(Locale.ROOT),
                  p = text(phone).trim(),
                  pw = text(pass);
              if (n.length() < 2) {
                name.setError("Họ tên phải có ít nhất 2 ký tự");
                return;
              }
              if (!Patterns.EMAIL_ADDRESS.matcher(e).matches()) {
                email.setError("Email không hợp lệ");
                return;
              }
              if (!p.matches("^[0-9+]{8,15}$")) {
                phone.setError("Số điện thoại không hợp lệ");
                return;
              }
              if (!pw.isEmpty() && pw.length() < 8) {
                pass.setError("Mật khẩu phải có ít nhất 8 ký tự");
                return;
              }
              AppExecutors.io()
                  .execute(
                      () -> {
                        try {
                          User current = repository.getUserById(Long.parseLong(data.getId()));
                          User dupeEmail = repository.getUserByEmail(e),
                              dupePhone = repository.getUserByPhone(p);
                          if (dupeEmail != null && dupeEmail.getId() != current.getId()) {
                            runOnUiThread(() -> email.setError("Email đã được sử dụng"));
                            return;
                          }
                          if (dupePhone != null && dupePhone.getId() != current.getId()) {
                            runOnUiThread(() -> phone.setError("Số điện thoại đã được sử dụng"));
                            return;
                          }
                          repository.updateUser(
                              current.updated(n, e, p, pw.isEmpty() ? current.getPassword() : pw));
                          if (!data.getEmail().equalsIgnoreCase(e))
                            RateLimiter.reset(this, data.getEmail());
                          boolean wasLocked = Boolean.TRUE.equals(data.getLocked());
                          if (locked.isChecked()) {
                            RateLimiter.lock(this, e);
                            if (!wasLocked)
                              repository.insertNotification(
                                  new AppNotification(
                                      0,
                                      current.getId(),
                                      "account_locked_" + System.currentTimeMillis(),
                                      "Tài khoản đã bị khóa",
                                      "Tài khoản RoomGo của bạn đã bị quản trị viên khóa.",
                                      "account",
                                      null,
                                      null,
                                      false,
                                      System.currentTimeMillis()));
                          } else {
                            RateLimiter.reset(this, e);
                          }
                          runOnUiThread(
                              () -> {
                                dialog.dismiss();
                                toast("Cập nhật người dùng thành công");
                                load();
                              });
                        } catch (Exception ex) {
                          runOnUiThread(() -> toast("Không thể cập nhật: " + ex.getMessage()));
                        }
                      });
            });
    dialog.show();
  }

  private void confirmDelete(AdminUserData u) {
    new MaterialAlertDialogBuilder(this)
        .setTitle("Xóa user")
        .setMessage("Bạn có chắc muốn xóa \"" + u.getFullName() + "\"?")
        .setPositiveButton(
            "Xóa",
            (d, w) ->
                AppExecutors.io()
                    .execute(
                        () -> {
                          User value = repository.getUserById(Long.parseLong(u.getId()));
                          if (value != null && repository.userHasBookings(value.getId())) {
                            runOnUiThread(() -> toast("Không thể xóa người dùng đang có booking"));
                            return;
                          }
                          if (value != null) repository.deleteUser(value);
                          runOnUiThread(
                              () -> {
                                adapter.removeUser(u.getId());
                                toast("Xóa user thành công!");
                              });
                        }))
        .setNegativeButton("Hủy", null)
        .show();
  }

  private static String text(TextInputEditText f) {
    return f.getText() == null ? "" : f.getText().toString();
  }

  private void toast(String s) {
    Toast.makeText(this, s, Toast.LENGTH_SHORT).show();
  }
}

File: app/src/main/java/com/example/homestay/AdminUsersActivity.kt
package com.example.homestay

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.lifecycleScope
import androidx.core.widget.addTextChangedListener
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.chip.ChipGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.homestay.data.model.AdminUserData
import com.example.homestay.ui.admin.AdminUserAdapter
import com.example.homestay.utils.RateLimiter
import com.example.homestay.utils.AdminAuth
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.first

class AdminUsersActivity : AppCompatActivity() {
    private val repository by lazy { (application as HomestayApplication).repository }
    
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: AdminUserAdapter
    private lateinit var progressBar: ProgressBar
    private var users = mutableListOf<AdminUserData>()
    private var lockedFilter: Boolean? = null
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_admin_users)
        
        setupToolbar()
        setupViews()
        loadUsers()
    }
    
    private fun setupToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = "Danh sách Users"
    }
    
    private fun setupViews() {
        recyclerView = findViewById(R.id.rv_users)
        progressBar = findViewById(R.id.progress_bar)
        
        adapter = AdminUserAdapter(
            users = users,
            onDeleteClick = { user -> showDeleteConfirmDialog(user) },
            onUnlockClick = { user -> showUnlockConfirmDialog(user) },
            onDetailsClick = { user -> showUserDetails(user) },
            onEditClick = { user -> showEditUserDialog(user) }
        )
        
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter
        findViewById<TextInputEditText>(R.id.et_admin_search).addTextChangedListener { filterUsers(it?.toString().orEmpty()) }
        findViewById<ChipGroup>(R.id.chip_user_status).setOnCheckedStateChangeListener { _, ids ->
            lockedFilter = when (ids.firstOrNull()) {
                R.id.chip_users_active -> false; R.id.chip_users_locked -> true; else -> null
            }
            filterUsers(findViewById<TextInputEditText>(R.id.et_admin_search).text?.toString().orEmpty())
        }
    }
    
    private fun loadUsers() {
        progressBar.visibility = View.VISIBLE
        lifecycleScope.launch {
            try {
                val allBookings = repository.getAllBookings().first()
                val usersList = repository.getAllUsers().first().map { user ->
                    val userBookings = allBookings.filter { it.userId == user.id }
                    val canLogin = RateLimiter.canAttemptLogin(this@AdminUsersActivity, user.email)
                    val failed = RateLimiter.getFailedAttempts(this@AdminUsersActivity, user.email)
                    AdminUserData(user.id.toString(), user.email, user.phone, user.fullName, user.createdAt,
                        failedLoginAttempts = failed, locked = !canLogin.first,
                        permanent = !canLogin.first && RateLimiter.getLockedSecondsRemaining(this@AdminUsersActivity, user.email) > 365L * 24 * 3600,
                        lockedUntil = canLogin.second,
                        secondsRemaining = RateLimiter.getLockedSecondsRemaining(this@AdminUsersActivity, user.email),
                        bookingCount = userBookings.size,
                        totalSpent = userBookings.filter { it.status == "completed" }.sumOf { it.totalPrice },
                        lastBookingAt = userBookings.maxOfOrNull { it.createdAt })
                }
                users.clear()
                users.addAll(usersList)
                filterUsers(findViewById<TextInputEditText>(R.id.et_admin_search).text?.toString().orEmpty())
            } catch (e: Exception) {
                android.util.Log.e("AdminUsers", "Error: ${e.message}", e)
                Toast.makeText(this@AdminUsersActivity, "Lỗi: ${e.message}", Toast.LENGTH_SHORT).show()
            } finally {
                progressBar.visibility = View.GONE
            }
        }
    }

    private fun filterUsers(query: String) {
        val keyword = query.trim()
        val filtered = users.filter {
            (lockedFilter == null || it.locked == lockedFilter) &&
                (it.fullName.contains(keyword, true) || it.email.contains(keyword, true) || it.phone.contains(keyword, true))
        }
        adapter.updateUsers(filtered)
        findViewById<View>(R.id.tv_empty).visibility = if (filtered.isEmpty()) View.VISIBLE else View.GONE
    }
    
    private fun showDeleteConfirmDialog(user: AdminUserData) {
        MaterialAlertDialogBuilder(this)
            .setTitle("Xóa user")
            .setMessage("Bạn có chắc chắn muốn xóa user \"${user.fullName}\" (${user.email})?")
            .setPositiveButton("Xóa") { _, _ ->
                deleteUser(user.id)
            }
            .setNegativeButton("Hủy", null)
            .show()
    }

    private fun showUserDetails(user: AdminUserData) {
        lifecycleScope.launch {
            val bookings = repository.getAllBookings().first().filter { it.userId.toString() == user.id }
            val rooms = repository.getAllRooms().first().associateBy { it.id }
            val date = java.text.SimpleDateFormat("dd/MM/yyyy", java.util.Locale("vi", "VN"))
            val history = bookings.sortedByDescending { it.createdAt }.take(10).joinToString("\n") {
                "• ${rooms[it.roomId]?.name ?: "Phòng #${it.roomId}"} — ${date.format(java.util.Date(it.checkInDate))} — ${it.status}"
            }.ifBlank { "Chưa có lịch sử đặt phòng" }
            val spent = java.text.NumberFormat.getNumberInstance(java.util.Locale("vi", "VN")).format(user.totalSpent.toLong())
            MaterialAlertDialogBuilder(this@AdminUsersActivity)
                .setTitle(user.fullName)
                .setMessage("Mã người dùng: ${user.id}\nEmail: ${user.email}\nĐiện thoại: ${user.phone}\nNgày tham gia: ${date.format(java.util.Date(user.createdAt))}\nTrạng thái: ${if (user.locked == true) "Bị khóa" else "Hoạt động"}\nTổng booking: ${user.bookingCount}\nTổng đã chi: $spent đ\n\nLịch sử gần đây:\n$history")
                .setPositiveButton("Đóng", null).show()
        }
    }

    private fun showEditUserDialog(userData: AdminUserData) {
        if (userData.email.equals(AdminAuth.EMAIL, ignoreCase = true)) {
            Toast.makeText(this, "Không thể chỉnh sửa tài khoản quản trị hệ thống", Toast.LENGTH_SHORT).show()
            return
        }

        val view = layoutInflater.inflate(R.layout.dialog_admin_edit_user, null)
        val etName = view.findViewById<TextInputEditText>(R.id.et_admin_user_name)
        val etEmail = view.findViewById<TextInputEditText>(R.id.et_admin_user_email)
        val etPhone = view.findViewById<TextInputEditText>(R.id.et_admin_user_phone)
        val etPassword = view.findViewById<TextInputEditText>(R.id.et_admin_user_password)
        val btnCancel = view.findViewById<com.google.android.material.button.MaterialButton>(R.id.btn_cancel)
        val btnSave = view.findViewById<com.google.android.material.button.MaterialButton>(R.id.btn_save)
        etName.setText(userData.fullName)
        etEmail.setText(userData.email)
        etPhone.setText(userData.phone)

        val dialog = MaterialAlertDialogBuilder(this).setView(view).create()
        btnCancel.setOnClickListener { dialog.dismiss() }
        btnSave.setOnClickListener {
            val name = etName.text?.toString()?.trim().orEmpty()
            val email = etEmail.text?.toString()?.trim()?.lowercase().orEmpty()
            val phone = etPhone.text?.toString()?.trim().orEmpty()
            val newPassword = etPassword.text?.toString().orEmpty()

            when {
                name.length < 2 -> etName.error = "Họ tên phải có ít nhất 2 ký tự"
                !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches() -> etEmail.error = "Email không hợp lệ"
                !phone.matches(Regex("^[0-9+]{8,15}$")) -> etPhone.error = "Số điện thoại không hợp lệ"
                newPassword.isNotBlank() && newPassword.length < 8 -> etPassword.error = "Mật khẩu phải có ít nhất 8 ký tự"
                else -> lifecycleScope.launch {
                    try {
                        val current = repository.getUserById(userData.id.toLong()) ?: return@launch
                        val duplicateEmail = repository.getUserByEmail(email)?.takeIf { it.id != current.id }
                        val duplicatePhone = repository.getUserByPhone(phone)?.takeIf { it.id != current.id }
                        when {
                            duplicateEmail != null -> etEmail.error = "Email đã được sử dụng"
                            duplicatePhone != null -> etPhone.error = "Số điện thoại đã được sử dụng"
                            else -> {
                                repository.updateUser(current.copy(
                                    fullName = name, email = email, phone = phone,
                                    password = newPassword.ifBlank { current.password }
                                ))
                                RateLimiter.reset(this@AdminUsersActivity, userData.email)
                                Toast.makeText(this@AdminUsersActivity, "Cập nhật người dùng thành công", Toast.LENGTH_SHORT).show()
                                dialog.dismiss()
                                loadUsers()
                            }
                        }
                    } catch (e: Exception) {
                        Toast.makeText(this@AdminUsersActivity, "Không thể cập nhật: ${e.message}", Toast.LENGTH_LONG).show()
                    }
                }
            }
        }
        dialog.show()
    }
    
    private fun showUnlockConfirmDialog(user: AdminUserData) {
        MaterialAlertDialogBuilder(this)
            .setTitle("Mở khóa tài khoản")
            .setMessage("Bạn có chắc chắn muốn mở khóa tài khoản \"${user.fullName}\" (${user.email})?")
            .setPositiveButton("Mở khóa") { _, _ ->
                unlockUser(user.id)
            }
            .setNegativeButton("Hủy", null)
            .show()
    }
    
    private fun unlockUser(userId: String) {
        // Tìm user để lấy email trước khi unlock
        val user = users.find { it.id == userId }
        val userEmail = user?.email
        
        progressBar.visibility = View.VISIBLE
        lifecycleScope.launch {
            userEmail?.let { RateLimiter.reset(this@AdminUsersActivity, it) }
            Toast.makeText(this@AdminUsersActivity, "Mở khóa tài khoản thành công!", Toast.LENGTH_SHORT).show()
            progressBar.visibility = View.GONE
            loadUsers()
        }
    }

    private fun deleteUser(userId: String) {
        lifecycleScope.launch {
            try {
                val user = repository.getUserById(userId.toLong()) ?: return@launch
                if (repository.userHasBookings(user.id)) {
                    Toast.makeText(this@AdminUsersActivity, "Không thể xóa người dùng đang có booking", Toast.LENGTH_LONG).show()
                    return@launch
                }
                repository.deleteUser(user)
                Toast.makeText(this@AdminUsersActivity, "Xóa user thành công!", Toast.LENGTH_SHORT).show()
                adapter.removeUser(userId)
            } catch (e: Exception) {
                Toast.makeText(this@AdminUsersActivity, "Lỗi: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }
    
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}
